### ISSUE TITLE
False positive with Box<dyn Iterator<>> and iter_not_returning_iterator 

### ISSUE THREAD

**Issue by HollayHorvath** [rust-lang/rust-clippy#8308]
### Summary

Having an function returning a boxed iterator triggers this lint.

### Lint Name

iter_not_returning_iterator 

### Reproducer

I tried this code:

```rust
pub trait Foo {
    fn iter(&self) -> Box<dyn Iterator<Item = i32> + '_>;
}
```

I saw this happen:

```
warning: this method is named `iter` but its return type does not implement `Iterator`
 --> src/lib.rs:3:5
  |
3 |     fn iter(&self) -> Box<dyn Iterator<Item = i32> + '_>;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: the lint level is defined here
 --> src/lib.rs:1:9
  |
1 | #![warn(clippy::iter_not_returning_iterator)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#iter_not_returning_iterator
```

I expected to see this happen:
No warn being triggered

### Version

```text
rustc 1.60.0-nightly (bd3cb5256 2022-01-16)
binary: rustc
commit-hash: bd3cb52565faab2755ff1bdb54d88bc91f47b4b9
commit-date: 2022-01-16
host: x86_64-unknown-linux-gnu
release: 1.60.0-nightly
LLVM version: 13.0.0
```

**Comment by Jarcho** [comment]
Fixed in #8287

**Comment by HollayHorvath** [comment]
I'm getting the same error with the latest clippy version (clippy 0.1.60 (ee5d8d3 2022-01-17)). Is it possible that #8287 isn't landed yet on nightly?

**Comment by xFrednet** [comment]
Correct, it takes some time until Clippy changes are distributed with nightly. Tomorrow's version should have this fix as it was synced and merged today in https://github.com/rust-lang/rust/pull/93001 :upside_down_face:
